pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java

```diff
@@ -1854,23 +1859,63 @@ private void setRowBufferColumn(Tuple rowBuffer,
         // or we won't be able to re-parse it.
         //
         case Types.DATE:
-          rowBuffer.set(columnIndex, connection
-              .encodeString(
-                  connection.getTimestampUtils().toString(
-                      getDefaultCalendar(), (Date) valueObject)));
+          if (valueObject instanceof Timestamp) {
+            rowBuffer.set(columnIndex, connection
+                .encodeString(
+                    connection.getTimestampUtils().toString(
+                        getDefaultCalendar(), (Date) valueObject)));
+
+          } else if (valueObject instanceof LocalDate) {
+            rowBuffer.set(columnIndex, connection
+                .encodeString(
+                    connection.getTimestampUtils().toString(
+                        getDefaultCalendar(), Date.valueOf((LocalDate) valueObject))));
+          } else {
+            throw new PSQLException(GT.tr("conversion to {0} from {1} not supported",
+                Types.DATE, valueObject.getClass().getName()),
+                PSQLState.INVALID_PARAMETER_VALUE);
+          }
           break;
 
         case Types.TIME:
-          rowBuffer.set(columnIndex, connection
-              .encodeString(
-                  connection.getTimestampUtils().toString(
-                      getDefaultCalendar(), (Time) valueObject)));
+          if (valueObject instanceof Time) {
+            rowBuffer.set(columnIndex, connection
+                .encodeString(
+                    connection.getTimestampUtils().toString(
+                        getDefaultCalendar(), (Time) valueObject)));
+
+          } else if (valueObject instanceof LocalTime) {
+            rowBuffer.set(columnIndex, connection
+                .encodeString(
+                    connection.getTimestampUtils().toString(
+                        getDefaultCalendar(), Time.valueOf((LocalTime) valueObject))));
+          } else {
+            throw new PSQLException(GT.tr("conversion to {0} from {1} not supported",
+                Types.TIME, valueObject.getClass().getName()),
+                PSQLState.INVALID_PARAMETER_VALUE);
+          }
           break;
 
         case Types.TIMESTAMP:
-          rowBuffer.set(columnIndex, connection.encodeString(
-              connection.getTimestampUtils().toString(
-                  getDefaultCalendar(), (Timestamp) valueObject)));
+          if (valueObject instanceof Timestamp) {
+            rowBuffer.set(columnIndex, connection.encodeString(
+                connection.getTimestampUtils().toString(
+                    getDefaultCalendar(), (Timestamp) valueObject)));
+
+          } else if (valueObject instanceof LocalDateTime) {
+            rowBuffer.set(columnIndex, connection.encodeString(
```

Comment:
Why aren't you using `org.postgresql.jdbc.TimestampUtils#toString(LocalDateTime)` here? Going through `Timestamp` likely leads to issues with DST transitions.

Reply:
Thanks, fixed.